### PR TITLE
Changes in views plugin

### PR DESCRIPTION
**Pull request by Cookiezaurs**
[SERVER-172] [SERVER-684] [SERVER-992]
Changes:
-- views plugin - using server side table, changed the way we store data(to get calculated values fast for a lot of views), Added average scroll calculation
--countly.helpers.js - small fix to don't allow hiding table columns without title, to load scrolling propertly, to have right numbers in description(when some columns without title)
--fetch.js - added function to fetch for views plugin. Modified to don't select from splitted documents if we know they are not splitted
--dbviewer plugin - modified to show changed views plugin collections correctly
-- data migration plugin --- added new collections to export

### CODE REVIEW

**Comment by ar2rsawseen** [comment]
This pull request **introduces 14 alerts** when merging 07cbe4d8a49f572f1a57e650c9605b3716762709 into ab8136c0a990a8526a79d657ff24d738888b4ffb - [view on LGTM.com](https://lgtm.com/projects/g/Countly/countly-server/rev/pr-aff31c7c4d37216f73ea418d6baa4fcedcc20748)

**new alerts:**

* 10 for Unused variable, import, function or class
* 2 for Variable not declared before use
* 1 for Useless conditional
* 1 for Missing variable declaration

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by ar2rsawseen** [comment]
This pull request **introduces 2 alerts** when merging 9c4edff2e6d3d366cc171ae193691b7df111e7e9 into ab8136c0a990a8526a79d657ff24d738888b4ffb - [view on LGTM.com](https://lgtm.com/projects/g/Countly/countly-server/rev/pr-b4dd7ae01985a302837e5b1f0e41ae0812917d3d)

**new alerts:**

* 1 for Useless conditional
* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: bin/upgrade/19.01/scripts/updateViews.js [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+                            catch(e){}
+                            for(var d in summedZero['no-segment']) { delete summedZero['no-segment'][d];}
+                            for(var d in summedZero['platform']) { delete summedZero['platform'][d];}
+    
+    }
+}
+
+function fixDocuments(retry, appID, done) {
+    console.log("Transforming views info");
+    segments = {};
```
[the remark this answers is not in the document]
Comment (can-keklik) [reply bin/upgrade/19.01/scripts/updateViews.js]: This file may need a lint check.

File: bin/upgrade/19.01/scripts/updateViews.js [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+        }
+    });
+}
+
+//Fixing data about users
+function check_and_fix_data(appID, done) {
+    viewsMap ={};
+    console.log("Updating for  view: "+appID);
+    console.log("Getting all view names");
+    countlyDb.collection('app_viewdata' + appID).aggregate([{$group:{_id:false,views:{$mergeObjects: "$meta_v2.views"}}}],function(err,res) {
```
[the remark this answers is not in the document]
Comment (can-keklik) [reply bin/upgrade/19.01/scripts/updateViews.js]: Encountered in "plugins/views/api/api.js" too.

File: api/parts/data/fetch.js [hunk 3]
```diff
@@ -685,6 +685,18 @@ fetch.fetchLoyalty = function(params) {
...
+* @param {string} collection -collection name
+* @param {params} params - params object
+* @param {boolean} isCustomEvent - if is custom event
+* @param {object} options - options object
+* @param {function} callback - calback function
+**/
+fetch.fetchViews = function(collection, params, isCustomEvent, options, callback) {
+    fetchTimeObj(collection, params, isCustomEvent, options, callback);
+};
+
```
Comment (ar2rsawseen) [inline api/parts/data/fetch.js]: Let's not add plugins related methods in core
better use this:
https://github.com/Countly/countly-server/blob/master/api/parts/data/fetch.js#L915-L919

File: frontend/express/public/javascripts/countly/countly.helpers.js [hunk 4]
```diff
@@ -1151,52 +1153,74 @@
...
-        dtable.CoultyColumnSel.maxCol = maxCol;
+
         dtable.CoultyColumnSel.tableCol = tableCols.length;
 
         var str = "";
         var myClass = "";
         var myClass2 = "";
         var disabled = "";
-        var selectedC = tableCols.length;
+        var selectedC = Math.min(maxCol);
```
Comment (ar2rsawseen) [inline frontend/express/public/javascripts/countly/countly.helpers.js]: No need for min

**Comment by ar2rsawseen** [comment]
This pull request **introduces 2 alerts** when merging 0010b1248325bd5d7bd01300591d62590ed3ca53 into 7ff15c69dd073a691fecd53d02a1cd116fdb0001 - [view on LGTM.com](https://lgtm.com/projects/g/Countly/countly-server/rev/pr-bfc14bf96bfd30b08592138b6e0f917eb38e20b7)

**new alerts:**

* 1 for Useless conditional
* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: plugins/views/api/api.js [hunk 5]
```diff
@@ -39,122 +44,683 @@ var pluginOb = {},
         var params = ob.params;
         var appId = params.qstring.app_id;
         var url = params.qstring.view_url;
-        var encodedUrl = common.db.encode(url);
+        var viewid = params.qstring.view_id;
+        //var encodedUrl = common.db.encode(url);
```
Comment (ar2rsawseen) [inline plugins/views/api/api.js]: Test with view name and/or url with dots, like dashboard.html

File: plugins/views/api/api.js [hunk 6]
```diff
@@ -3,16 +3,21 @@ var pluginOb = {},
     request = require('request'),
     Promise = require("bluebird"),
     common = require('../../../api/utils/common.js'),
+    moment = require('moment-timezone'),
     authorize = require('../../../api/utils/authorizer.js'),
     countlyCommon = require('../../../api/lib/countly.common.js'),
     plugins = require('../../pluginManager.js'),
     fetch = require('../../../api/parts/data/fetch.js'),
-    countlyModel = require('../../../api/lib/countly.model.js');
```
Comment (ar2rsawseen) [inline plugins/views/api/api.js]: Need to check if someone uses server side model file and if it is up to date and still working.
https://github.com/Countly/countly-server/blob/master/plugins/views/api/lib/countly.model.js

**Comment by ar2rsawseen** [comment]
This pull request **introduces 2 alerts** when merging 90e91161e9b4067cdc84c635387b2562835cb319 into 8fa23ededc6967d69d825aee1a83ec6a01875ff5 - [view on LGTM.com](https://lgtm.com/projects/g/Countly/countly-server/rev/pr-78b8f1582aa12f9544be19819d22bfae758dc2d5)

**new alerts:**

* 1 for Useless conditional
* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by ar2rsawseen** [comment]
This pull request **introduces 2 alerts** when merging b869706c765989cab3384755b8794320e2c59814 into 8fa23ededc6967d69d825aee1a83ec6a01875ff5 - [view on LGTM.com](https://lgtm.com/projects/g/Countly/countly-server/rev/pr-7e959a1a1353d1113e7957b19316073a9fe90f70)

**new alerts:**

* 1 for Useless conditional
* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**